File: C/remove_invalid_paranthesis.c
// https://www.geeksforgeeks.org/remove-invalid-parentheses/
/*
An expression will be given which can contain open and close parentheses 
and optionally some characters, No other operator will be there in string. 
We need to remove minimum number of parentheses to make the input string valid. 

If more than one valid output are possible removing same number of parentheses 
then print all such output.

Examples:
Input  : str = “()())()” -
Output : ()()() (())()

Input  : str = (v)())()
Output : (v)()()  (v())()
*/

// Python based sol using que @remove_invalid_paranthesis.py
// Below sol. can print same valid exp. more than once. In python this is 
// avoided using set.

#include <stdio.h>
#include <limits.h>
#include <malloc.h>
#include <string.h>

char* copy_arr(char *ip, int length){
    char *op = (char*)calloc(length+1,sizeof(char));
    int i = 0;
    for(;i<length;i++){
        op[i] = ip[i];
    }
    op[i] = '\0';
    return op;
}

typedef struct node{
    char *data;
    int length;
    struct node *next;
}list_node;

list_node* make_node(char *data, int length){
    list_node *node = (list_node*)malloc(sizeof(list_node));
    if(node){
        node->data = copy_arr(data,length);
        node->length = length;
        node->next = node;
    }
    return node;
}

list_node* append_node(list_node *last, list_node *data){
    if(data==NULL){
        return last;
    }else{
        if(last){
            data->next = last->next;
            last->next = data;
        }
        return data;
    }
}

void print_list(list_node *last){
    if(last){
        list_node *first = last->next;
        for(;first!=last;first=first->next){
            printf("%s ",first->data);
        }
        printf("%s\n",first->data);
    }
}

typedef struct {
    int min_bracket_removed;
    list_node *last;
}valid_list;

valid_list* init_valid_list(){
    valid_list *list = (valid_list*)malloc(sizeof(valid_list));
    if(list){
        list->min_bracket_removed = INT_MAX;
        list->last = NULL;
    }
    return list;
}

int is_paranthesis(char c){
    if(c=='(' || c==')') return 1;
    return 0;
}

int is_valid_exp(char *ip, int length){
    int count = 0;
    int i = 0;
    for(;i<length;i++){
        if(ip[i]=='(') count++;
        else if(ip[i]==')') count--;
        if(count<0) return 0;
    }
    if(count>0) return 0;
    return 1;
}

void print_valid_expression(char *ip, int length, int index, 
    int bracket_removed_count, char *op, int k, valid_list *list){
        if(length<0 || index <0 || k<0) return;
        if(index==length){
            if(is_valid_exp(op,k) && 
                bracket_removed_count<=list->min_bracket_removed){
                    if( list->last && 
                        bracket_removed_count<list->min_bracket_removed){
                        free(list->last->next);
                        list->last = NULL;
                        list->min_bracket_removed = INT_MAX;
                    }
                    list->last = append_node(list->last,make_node(op,k));
                    list->min_bracket_removed = bracket_removed_count;
                }
            return;
        }
        op[k] = ip[index];
        print_valid_expression(ip,length,index+1,bracket_removed_count,op,k+1,
            list);
        if(is_paranthesis(ip[index]) && 
            bracket_removed_count<list->min_bracket_removed){
            print_valid_expression(ip,length,index+1,bracket_removed_count+1,op,
                k,list);
        }
}

int main(){
    char *ip = "(v)(((v)";
    int length = strlen(ip);

    printf("Ip=>%s\n",ip);

    char *op = (char*)calloc(length+1,sizeof(char));
    valid_list *list = init_valid_list();
    print_valid_expression(ip,length,0,0,op,0,list);
    printf("Min. bracet removed = %d\n",list->min_bracket_removed);
    print_list(list->last);

    return 0;
}

File: Python/remove_invalid_paranthesis.py
# https://www.geeksforgeeks.org/remove-invalid-parentheses/
"""
An expression will be given which can contain open and close parentheses 
and optionally some characters, No other operator will be there in string. 
We need to remove minimum number of parentheses to make the input string valid. 

If more than one valid output are possible removing same number of parentheses 
then print all such output.

Examples:
Input  : str = “()())()” -
Output : ()()() (())()

Input  : str = (v)())()
Output : (v)()()  (v())()
"""
# C based sol using que @remove_invalid_paranthesis.c

from collections import deque

def is_valid_exp(ip_str):
    count = 0
    for char in ip_str:
        if char=='(':
            count+=1
        elif char==')':
            count-=1
        if count<0:
            return False
    if count>0:
        return False
    return True

def is_paranthesis(c):
    if c == ')' or c =='(':
        return True
    return False

def check_and_insert(unique_exp_set, que, ip_str):
    if not unique_exp_set.__contains__(ip_str):
        unique_exp_set.add(ip_str)
        que.append(ip_str)

def get_valid_exp(ip_str):
    min_bracket_removed = 0
    result_list = list()
    unique_exp_set = set()
    que = deque()
    null_str = 'NULL'
    check_and_insert(unique_exp_set,que,ip_str)
    que.append(null_str)
    got_valid_exp = False
    while not len(que)==0 and (que[0] is not null_str):
        print(que)
        while not len(que)==0 and (que[0] is not null_str):
            temp = que.popleft()
            if is_valid_exp(temp):
                got_valid_exp = True
                result_list.append(temp)
            else:
                for i in range(len(temp)):
                    if not is_paranthesis(temp[i]):
                        continue
                    next_exp = temp[0:i] + temp[i+1:len(temp)]
                    check_and_insert(unique_exp_set,que,next_exp)
        que.popleft()
        min_bracket_removed+=1
        que.append(null_str)
        if got_valid_exp:
            return min_bracket_removed-1, result_list
    return min_bracket_removed-1, result_list

def main():
    ip_str = '(v)(((v)'
    print('Ip=>',ip_str)
    min_bracket_removed, result_list = get_valid_exp(ip_str)
    print('Min. bracket removed =',min_bracket_removed)
    print(result_list)

if __name__=='__main__':
    main()